[Experimental] Refact Rating filter block for the new Product Filters block structure

retrofox opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

This PR updates the **Rating Filter** block to work correctly with the new structure of the **Products Filter**.

### Disclaimer

This RP does not handle the Dropdown mode of the filter. I suggest to table it in a follow-up PR

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #51639

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

0. Ensure the experimental blocks are enabled:
    - If you use the WooCommerce Beta Tester plugin, you can enable the experimental blocks in Tools > WCA Test Helper > Features. Toggle the `experimental-blocks` option.
    - If you build this PR locally, build the branch with `pnpm --filter='@woocommerce/plugin-woocommerce' build` command to have experimental blocks available.
1. Add the **Product Filters (Experimental)** block to the Product Catalog template.
2. Save and go to the shop page.
4. Select some filters, see filter blocks react, and update the filter options accordingly.
5. Interact with the **Rating** filter slider, see the filter updated accordingly.

https://github.com/user-attachments/assets/5d42339a-130d-4274-b9ad-7953943a7d1d

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [x] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

retrofox committed:
update Rating experimental filter title

retrofox committed:
allow to insert the filter block

retrofox committed:
fix ancestor block name

retrofox committed:
remove showFilterButton from attribute types

retrofox committed:
udpate icon

retrofox committed:
add rating to the initial filters template

retrofox committed:
clean rating filter utils

retrofox committed:
minor styling code enhancement

retrofox committed:
tidy

retrofox committed:
remove isEditor const

retrofox committed:
import type explicitely

retrofox committed:
rename block to CheckboxListEdit

retrofox committed:
set defautl style for Rating block

retrofox committed:
rename block to RatingFilterEdit

retrofox committed:
undestructuring attributes object

retrofox committed:
compose rating block by using template

retrofox committed:
use InitialDisabled instead of Disabled

retrofox committed:
set block instance name

retrofox committed:
set display style properly

retrofox committed:
do not rename checkbox list block instance

retrofox committed:
passing context to CheckboxList block, server-side

retrofox committed:
import type explicitely

retrofox committed:
process rating data in server-side

retrofox committed:
handle activate/deactivate rating filters

retrofox committed:
support clean rating filters in the frontend

retrofox committed:
remove temporarely action to handle dropdown

retrofox committed:
changelog

retrofox committed:
remove Dropdown mode for now

dinhtungdu commented on plugins/woocommerce-blocks/assets/js/blocks/product-filters/inner-blocks/rating-filter/block.json:
```diff
@@ -1,22 +1,19 @@
-	"keywords": [
-		"WooCommerce"
-	],
+	"keywords": [],
```
Let's keep the keyword.

retrofox replied:
Thanks for the feedback!
Since the block is already in the `WooCommerce` category, is there a specific reason to keep it as a keyword?

dinhtungdu replied:
You're right. I didn't have a strong opinion, just keeping the consistency. But given that Rating appears in the search result of "woocommerce" without the keyword, I agree that we don't need it.

dinhtungdu commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductFilterRating.php:
```diff
@@ -114,49 +114,53 @@ protected function render( $attributes, $content, $block ) {
+		 * Get the rating items
+		 * based on the selected ratings and the rating counts.
+		 */
+		$items = $this->get_checkbox_list_items( $rating_counts, $selected_rating, $attributes['showCounts'] ?? false );
```
As we also support the chip styles, I think the method name here should be more generic, like `get_rating_items`.

dinhtungdu commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductFilterRating.php:
There are some unused imports and methods in this class that should be removed.

dinhtungdu commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductFilterRating.php:
```diff
@@ -114,49 +114,53 @@ protected function render( $attributes, $content, $block ) {
-			)
-		) : Dropdown::render(
-			$this->get_dropdown_props( $filtered_rating_counts, $selected_ratings_query_param, $show_counts, $attributes['selectType'] )
+		$display_style  = $attributes['displayStyle'] ?? 'list';
+		$show_counts    = $attributes['showCounts'] ?? false;
```
These variables are unused.

retrofox replied:
Same as before. It's part of the dropdown mode, if I'm not wrong. Are they triggering linting issues or something like that?

retrofox replied:
removed [here](https://github.com/woocommerce/woocommerce/pull/52152/commits/1b2d8c543cbc141637f52124573ed444c46730de)

dinhtungdu commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductFilterRating.php:
```diff
@@ -216,7 +218,7 @@ function( $rating ) use ( $ratings_array, $show_counts ) {
 				return array(
 					'id'         => 'rating-' . $rating_str,
-					'checked'    => in_array( $rating_str, $ratings_array, true ),
+					'selected'   => in_array( $rating_str, $selected_ratings_query, true ),
```
The `$selected_ratings_query` variable is declared as string.

retrofox replied:
addressed [here](https://github.com/woocommerce/woocommerce/pull/52152/commits/afc1266086088acc9de92144a300c7d73abd1584)

dinhtungdu commented on plugins/woocommerce-blocks/assets/js/blocks/product-filters/inner-blocks/rating-filter/edit.tsx:
```diff
@@ -14,19 +17,22 @@ import {
+import type { Attributes } from './types';
 
 const NoRatings = () => (
 	<Notice status="warning" isDismissible={ false }>
```
Let's use [this component](https://github.com/woocommerce/woocommerce/blob/update%2Fwoo-blocks-refact-rating-block/plugins/woocommerce-blocks/assets/js/blocks/product-filters/components/notice/index.tsx) instead. I shared the reason for it in the comment of that component.

retrofox replied:
This is a good example of why we use custom implementation for something already handled by core.

<img width="928" alt="image" src="https://github.com/user-attachments/assets/061ea7b6-fc7b-46f2-89d7-7cf1dbe0647c">

Why must we create a new abstraction to notify something we could consider a regular notice? Why does the notice for the product filter need to be different for the rest of the app?


The following is the jsdoc for the custom component:

```
/**
 * A custom notice component is designed specifically for new filter blocks. We
 * are not reusing the existing components because we have a new design for the
 * filter blocks notice. We want users to utilize the sidebar for attribute
 * settings, so we are keeping the new notice minimal."
 */
 ```

I don't get the _We want users to utilize the sidebar for attribute settings, so we are keeping the new notice minimal._ part. It seems that the component was designed to be used in the Sidebar?

Anyway, I don't see the reason for introducing a new abstraction for something that isn't part of the block visualization but the app UI.

retrofox replied:
btw I'm entirely okay to use the custom component.

dinhtungdu commented on plugins/woocommerce-blocks/assets/js/blocks/product-filters/inner-blocks/rating-filter/edit.tsx:
```diff
@@ -14,19 +17,22 @@ import {
 import { previewOptions } from './preview';
-import { Attributes } from './types';
 import { getActiveFilters } from './utils';
 import { useSetWraperVisibility } from '../../../filter-wrapper/context';
```
We don't need this anymore.

dinhtungdu replied:
Look like this was missed.

dinhtungdu replied:
You can remove it in a follow up PR along with Dropdown if you want.

retrofox replied:
let's do it in this one. Thanks for the note

retrofox replied:
Do you mean to the hook, right? removed [here](https://github.com/woocommerce/woocommerce/pull/52152/commits/a712a3f6a9c35b292240380f04e20a4c21a2939c)

dinhtungdu commented on plugins/woocommerce-blocks/assets/js/blocks/product-filters/inner-blocks/rating-filter/edit.tsx:
```diff
@@ -14,19 +17,22 @@ import {
 import { useSetWraperVisibility } from '../../../filter-wrapper/context';
 import { Inspector } from './components/inspector';
+import { InitialDisabled } from '../../components/initial-disabled';
 import { PreviewDropdown } from '../components/preview-dropdown';
```
We don't need this now, too.

retrofox replied:
`PreviewDropdown`? It's part of the Dropdown mode. Could we leave the code there and work on it in a follow-up?

dinhtungdu replied:
We don't have the Dropdown mode any more. But it make sense to remove it in a follow up PR.

dinhtungdu commented on plugins/woocommerce-blocks/assets/js/blocks/product-filters/inner-blocks/rating-filter/edit.tsx:
```diff
@@ -157,50 +208,44 @@ const Edit = ( props: BlockEditProps< Attributes > ) => {
-									}
-								/>
-							</>
+						{ displayStyle === 'dropdown' ? (
```
Display style now should be the full name of the inner block, like `woocommerce/product-filter-checkbox-list`. We always pass the filter options down through the block context. You can see Stock or Attribute for inspiration. The difference here is the Rating block only supports one display style.

retrofox replied:
nice to know. Could we tackle the dropdown mode in a follow-up? I noted that in the PR description, in the `Disclaimer` section

dinhtungdu replied:
As mentioned in another comment, let's remove it in a follow up PR.

dinhtungdu reviewed (changes requested):
Thanks for working on this PR. The filter worked as expected in my test. The display style part needs some updates to follow the current structure. Please refer to my review comments for more details.

roykho commented on plugins/woocommerce-blocks/assets/js/blocks/product-filters/inner-blocks/checkbox-list/edit.tsx:
```diff
@@ -205,4 +205,4 @@ export default withColors( {
 	optionElementSelected: 'option-element-border',
 	optionElement: 'option-element',
-} )( Edit );
+} )( CheckboxListEdit );
```
Should we keep to the convention and use Edit just like all other blocks?

retrofox replied:
setting the name for the component helps to debug. Easy to find the component in the dev tool

roykho commented on plugins/woocommerce/changelog/update-woo-blocks-refact-rating-block:
```diff
@@ -0,0 +1,4 @@
+Significance: patch
+Type: update
+
+Rating Filter block: refact block according to the new Product Filters block structure
```
So since we're not announcing this change (experimental) we should make this changelog entry a comment type.

retrofox replied:
Like the suggested commit above?

roykho replied:
So I meant a changelog like this is a comment type: https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/changelog/51227-add-delayed-account-creation-send-password-51193

roykho reviewed (changes requested):
Good job so far! In addition to what Tung mention, I have left some comments.

Also the ratings display is reversed in the frontend. It is showing lowest rating first when it should show highest rating first.

retrofox committed:
use custom Notice component

retrofox committed:
do not import unused libs

retrofox committed:
rename method to get_rating_items

retrofox committed:
declar selected_ratings_query as mixed

retrofox commented on plugins/woocommerce/changelog/update-woo-blocks-refact-rating-block:
```diff
@@ -0,0 +1,4 @@
+Significance: patch
+Type: update
```
```suggestion
Type: comment
```

retrofox commented:
> The display style part needs some updates to follow the current structure. 

Can you try to address the following issues in follow-ups? I'd like to try to keep the PR small.

> Please refer to my review comments for more details.

👍

retrofox commented:
> Also the ratings display is reversed in the frontend. It is showing lowest rating first when it should show highest rating first.

Nice one. The thing is that it [reverses the order in the client](https://github.com/woocommerce/woocommerce/blob/9e88dc92b2c14548b41a7c63d9422a8e786fcdba/plugins/woocommerce-blocks/assets/js/blocks/product-filters/inner-blocks/rating-filter/edit.tsx#L144) arbitrary, in the editor context.

I will reverse the array in the server, but I wonder whether we should find a consistent order criterion.
Even more, I wonder about having a block option for this.

retrofox committed:
sort rating stars by value descending

retrofox committed:
udpate changelog message

retrofox committed:
linting

dinhtungdu commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductFilterRating.php:
```diff
@@ -293,7 +293,20 @@ private function get_rating_counts( $block ) {
 		$counts = $filters->get_rating_counts( $query_vars );
-		$data   = array();
+
+		/*
+		 * Sort the counts by rating value in descending order.
+		 * Todo: Consider to implement this consistently in the query.
+		 * @see https://github.com/woocommerce/woocommerce/pull/52152#issuecomment-2437598323
+		 */
+		uksort(
+			$counts,
+			function ( $a, $b ) {
+				return $b - $a;
+			}
+		);
```
We should modify the query that returns that count itself. Since it's a simple change, can you please implement it in this PR?

See: https://github.com/woocommerce/woocommerce/blob/0e2b94567be835c9edacf3ccdeb114bfac129ad0/plugins/woocommerce/src/Blocks/QueryFilters.php#L151

retrofox replied:
yes, I think it does make sense.

retrofox replied:
Addressed [here](https://github.com/woocommerce/woocommerce/pull/52152/commits/7fe2076d9ba6cb9e11ad71de6421b3705c211830). Thanks for the tip.

dinhtungdu commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductFilterRating.php:
```diff
@@ -192,18 +194,16 @@ private function render_rating_label( $rating, $count_label ) {
-	 * @param string $selected_ratings_query The url query param for selected ratings.
-	 * @param bool   $show_counts      Whether to show the counts.
+	 * @param array $rating_counts          - The rating counts.
+	 * @param mixed $selected_ratings_query - The url query param for selected ratings.
```
Why this is `mixed` but not `array`?

dinhtungdu reviewed (changes requested):
Thanks for the update. This is mostly look good to me! Great work! I have a small issue that I occationally get the notice about empty rating when I do have rating. It happened after I insert the block to the template. I'm justing raising it here, feel free to address it in a follow up PR if needed, as this PR is already big.

<img width="581" alt="image" src="https://github.com/user-attachments/assets/34ecbba7-a8a4-46ef-b693-9e601ecc527d">

retrofox committed:
remove unused vars

retrofox committed:
set selected_ratings_query like an array

retrofox committed:
set get_rating_items $select param as an array

retrofox committed:
set explicitely Attributes type

retrofox committed:
remove unneeded InitialDisabled component

retrofox committed:
remove useSetWraperVisibility hook usage

retrofox committed:
sort rating straight via SQL
props to @dinhtungdu https://github.com/woocommerce/woocommerce/pull/52152/files#r1818839577

retrofox committed:
remove local state to show no products notice

retrofox commented:
> I have a small issue that I occationally get the notice about empty rating when I do have rating. It happened after I insert the block to the template. 

I think we've addressed the issue [in this commit](https://github.com/woocommerce/woocommerce/pull/52152/commits/99bbee8e17a557465b419389893a318ae73c0554).

Thanks a lot folks for your feedback. PR is ready for review again.

roykho reviewed (approved):
Thanks for the updates. Works great! Good job!

✅ The block is now in the shape of the new design structure
✅ The ratings display correctly in the editor and frontend
✅ The ratings filters the products correctly in the frontend
✅ The clear button works to clear all selected ratings

I think as a follow up to this PR after merging and create issues:
- Add the ability to show/hide the clear button
- Decide if we really need the dropdown option and create an issue to either remove all reminisce of the dropdown and implement it.
- Add color option for unselected stars vs selected.

dinhtungdu reviewed (approved):
Thanks for the update! Tested and still worked as expected. :shipit: 🎉

retrofox commented:
🎉 !!!!
